### PR TITLE
[XLA] Add back: lift bitcast PR 

### PR DESCRIPTION
**Pull request by nouiz**
#50153 was reverted.
I'm adding it back with a fix: now we do not lift if any instruction change the layout.
@akuegel

### CODE REVIEW

**Review by akuegel — APPROVED** [review]
Thanks :)

File: tensorflow/compiler/xla/service/gpu/BUILD
Comment (akuegel) [inline tensorflow/compiler/xla/service/gpu/BUILD:1233]: This target was removed yesterday, and depending on "//tensorflow/compiler/xla/service:hlo" is enough to get access to hlo_casting_utils.h header.
Can you please remove this line? This should also show up as compile error if you sync to HEAD.

**Review by akuegel — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/gpu/BUILD
Comment (nouiz) [reply tensorflow/compiler/xla/service/gpu/BUILD:1233]: I rebased and removed that line. Now it build and the test still pass.

**Review by akuegel — APPROVED** [review]

**Comment by nouiz** [comment]
I asked and the important is the user driver and the minimum driver version is 470 with CUDA 11.2.
So all should be good.
I added the printing of the driver version when the error happens. Just in case something strange make that test run with an older driver.

**Comment by gbaned** [comment]
@nouiz  Can you please resolve conflicts? Thanks!

**Comment by nouiz** [comment]
> @nouiz Can you please resolve conflicts? Thanks!

Done.

**Review by akuegel — APPROVED** [review]

**Comment by nouiz** [comment]
Any idea what is happening?

**Comment by akuegel** [comment]
> Any idea what is happening?

Sorry for the delay, the problem is that the process is complicated. I approved the PR in Gerrit when it was imported from Github, as such I was not allowed to also approve it when it was sent out internally for review (it has to be approved by someone else). Then, additionally the test we got from which I had extracted the reproducer did not work anymore by now, so I couldn't verify myself that the issue is completely fixed, and I warned the reporter of the issue that the changelist will land again. After they didn't react, @cheshire approved the changelist and it got committed.

**Comment by nouiz** [comment]
Thanks for the follow up.

**Comment by hawkinsp** [comment]
I bisected a new test failure in JAX to this PR. I've only seen the failure under CUDA 10.2. I was able to reproduce it on a T4 GP PU and a GTX 1080, so I don't think it's GPU-dependent.

The following HLO will reproduce the problem:
```
HloModule primitive_computation_reduce_sum.8

%primitive_computation_add.3 (parameter.4: f32[], parameter.5: f32[]) -> f32[] {
  %parameter.4 = f32[] parameter(0), metadata={op_type="add" op_name="add"}
  %parameter.5 = f32[] parameter(1), metadata={op_type="add" op_name="add"}
  ROOT %add.6 = f32[] add(f32[] %parameter.4, f32[] %parameter.5), metadata={op_type="add" op_name="add"}
}

ENTRY %primitive_computation_reduce_sum.8 (parameter.1: f32[2,20,25,4]) -> f32[4] {
  %parameter.1 = f32[2,20,25,4]{3,2,1,0} parameter(0), metadata={op_type="reduce_sum" op_name="reduce_sum[ axes=(0, 1, 2) ]"}
  %constant.2 = f32[] constant(0), metadata={op_type="reduce_sum" op_name="reduce_sum[ axes=(0, 1, 2) ]"}
  ROOT %reduce.7 = f32[4]{0} reduce(f32[2,20,25,4]{3,2,1,0} %parameter.1, f32[] %constant.2), dimensions={0,1,2}, to_apply=%primitive_computation_add.3, metadata={op_type="reduce_sum" op_name="reduce_sum[ axes=(0, 1, 2) ]"}
}
```

I'm going to attempt to roll back the PR.

(It's possible this is not a bug in this PR but perhaps in something else, e.g., the CUDA toolkit, but rolling back seems like the safest course for the moment.)

**Comment by nouiz** [comment]
What was the problem? Crash? wrong value? Something else?

**Comment by hawkinsp** [comment]
On device crash (`CUDA_ERROR_ILLEGAL_ADDRESS`).

**Comment by cheshire** [comment]
@nouiz Do we have a clear reproducer which shows that index calculations from the bitcast are too slow? I'm still really surprised it's non-negligible compared to the memory read. I'd also be very curious if it can be bypassed in other ways (e.g. using linear index directly and not decomposing it at every step)

**Comment by hawkinsp** [comment]
@nouiz I'm wondering if the problem I outlined above is a `ptxas` bug. I tried using a CUDA 10.2 installation but replacing the `ptxas` binary with the `ptxas` from CUDA 11.4. The test now passes.

If you can confirm this is a `ptxas` miscompilation that has already been fixed, we can probably just roll forward (given I doubt `ptxas` bugs in CUDA 10.2 are ever going to be fixed at this point...)?

**Comment by hawkinsp** [comment]
Edit: note: I just saw a very similar looking illegal address bug that also appears to involve a reduction (https://github.com/google/jax/issues/7752) and that also goes away with a newer `ptxas`.

**Comment by nouiz** [comment]
@hawkinsp I can't reproduce this on my current container.
I do not have an CUDA 10.2 container that works with upstream TF. Some libs are missing. Do you have one you can share?

@cheshire you undid recently a WAR again a CUDA bug. Could this be the issue? I'm talking about this revert commit:
366e91a4316874c93e6fc10307b1cda6c4f2f712

The commit said that it is fixed in CUDA 10.2. But I do not know if it is fixed for all CUDA10.2, or only the the latest update of CUDA 10.2.2. @hawkinsp Which CUDA 10.2.* do you use? Can you try CUDA 10.2.2 or revert commit 366e91a4316874c93e6fc10307b1cda6c4f2f712 and see if this fix the bug?

**Comment by cheshire** [comment]
> @cheshire you undid recently a WAR again a CUDA bug. Could this be the issue? I'm talking about this revert commit:
366e91a

Yes, it could be that one.

**Comment by nouiz** [comment]
> @nouiz Do we have a clear reproducer which shows that index calculations from the bitcast are too slow? I'm still really surprised it's non-negligible compared to the memory read. I'd also be very curious if it can be bypassed in other ways (e.g. using linear index directly and not decomposing it at every step)

I had a repro. I can dig it if you want. The issue was more pronounced on A100 then on V100.
Note, the indexing is not trivial indexing. Sometimes LLVM optimize it well, but not always.
When it doesn't, then we end up with division and module. XLA takes the linear index and map it back to an nd-indexing.
Integer division and modulo are costly. If you combine that with a fusion that does non-trivial computation, then you hit the problem.

Note, I already optimized such computation (linear indexing to nd-indexing) in previous project. So I'm not surprised to see that again.

I think for now, we should check if the issue is related to 366e91a or not.

**Comment by hawkinsp** [comment]
@nouiz Here's another way to reproduce: just put the `ptxas` from 10.2 in your `PATH` using a TF tree otherwise built using 11.x. (This means that XLA will use it to compile PTX; it's irrelevant what TF does for its kernels and this reproducer doesn't call anything precompiled.)

I'm using the CUDA debian packages; the `nvcc` package which contains `ptxas` reports itself as:
```
ii  cuda-nvcc-10-2                     10.2.89-1                                   amd64        CUDA nvcc
```

**Comment by hawkinsp** [comment]
I verified that locally reverting https://github.com/tensorflow/tensorflow/commit/366e91a4316874c93e6fc10307b1cda6c4f2f712 appears to fix the problem.

**Comment by nouiz** [comment]
So, I think the solution is to revert it and add back my PR?
Do I need to make a new PR?

I wasn't able to reproduce the bug.

**Comment by hawkinsp** [comment]
@nouiz Yes, I think that works. No need for a new PR, I'll just roll back the roll back I made.

**Comment by cheshire** [comment]
Maybe not rollback just yet, https://partners.nvidia.com/Bug/ViewBug/2827811 claims that the ptxas bug was fixed in 10.2, was it actually not fixed?

**Pull request merged**
